parsec/backend/asgi/rpc.py

```diff
@@ -42,56 +31,172 @@
 
 CONTENT_TYPE_MSGPACK = "application/msgpack"
 AUTHORIZATION_PARSEC_ED25519 = "PARSEC-SIGN-ED25519"
+SUPPORTED_API_VERSIONS = (
+    API_V2_VERSION,
+    API_V3_VERSION,
+)
 
 
-def rpc_msgpack_rep(data: dict) -> Response:
+rpc_bp = Blueprint("anonymous_api", __name__)
+
+
+def _rpc_msgpack_rep(data: dict, api_version: ApiVersion) -> Response:
     return Response(
         response=packb(data),
         # Unlike REST, RPC doesn't use status to encode operational result
         status=200,
         content_type=CONTENT_TYPE_MSGPACK,
+        headers={"Api-Version": str(api_version)},
     )
 
 
-rpc_bp = Blueprint("anonymous_api", __name__)
+# The HTTP RPC API is divided into two layers: handshake and actual command processing
+# The command processing is based on msgpack and is common with the Websocket API.
+#
+# On the other hand, the handshake part is specific to how works HTTP (i.e. the
+# handshake is done for each query in HTTP using headers, unlike the Websocket
+# handshake which is done once by sending challenge/reply messages).
+#
+# So the RPC API handshake cannot return a body in msgpack given at this level
+# we are not settled on what should be used yet (who knows ! maybe in the future
+# we will use another serialization format for the command processing).
+# Instead we rely on the following HTTP status code:
+# - 404: Organization non found or invalid organization ID
+# - 401: Bad authentication info (bad Author/Signature/Timestamp headers)
+# - 415: Bad content-type
+# - 422: Unsupported API version
 
 
-@rpc_bp.route("/anonymous/<raw_organization_id>", methods=["GET", "POST"])
-async def anonymous_api(raw_organization_id: str):
-    backend: BackendApp = g.backend
+def _handshake_abort(status_code: int, api_version: ApiVersion) -> NoReturn:
+    abort(Response(response="", status=status_code, headers={"Api-Version": str(api_version)}))
+
+
+async def _do_handshake(
+    raw_organization_id: str,
+    backend: BackendApp,
+    allow_missing_organization: bool,
+    check_authentication: bool,
+) -> Tuple[ApiVersion, OrganizationID, Optional[Organization], Optional[User], Optional[Device]]:
+    # The anynimous HTTP API existed before the `Api-Version`/`Content-Type` fields
+    # where introduced, hence we have this workaround to provide backward compatibility
+    request.headers.setdefault("Api-Version", "3.0")
+    request.headers.setdefault("Content-Type", CONTENT_TYPE_MSGPACK)
```

Comment:
Do we also want to add those headers in authenticated api ?
Maybe move them to `anonymous_api` ?

Reply:
At first I only patched the anonymous api, but it was cumbersome to test (special cases are never fun to work with !)

So I consider having the authenticated route slightly more forgiving than needed was a good tradeoff given the code&test simplification ;-)